### PR TITLE
Fix and simplify handling of options for MSVC

### PR DESCRIPTION
**Pull request by llunak**
These are various fixes (on top) of #506 .

### CODE REVIEW

File: src/argprocessing.cpp [hunk 1]
```diff
@@ -242,6 +242,13 @@ process_arg(const Context& ctx,
     return nullopt;
   }
 
+  if (ctx.config.compiler_type() == CompilerType::cl
+      && util::starts_with(args[i], "/")) {
+    // MSVC understands both /option and -option, so convert all /option to
+    // -option to simplify our handling.
+    args[i][0] = '-';
```
Comment (rvogg) [inline src/argprocessing.cpp:249]: With this change, you break absolute paths without drive letters. Depending on what build system you are using, it is quite possible that a posix style path is used without a drive letter. 
Also, the feature should work when cross compiling from Linux to Windows with clang-cl.

Comment (llunak) [reply src/argprocessing.cpp:249]: That doesn't seem to be the case:
C:\lo\ccache.git\build>cl -nologo -c /lo/ccache.git/build/a.cpp
cl : Command line warning D9002 : ignoring unknown option '/lo/ccache.git/build/a.cpp'
cl : Command line error D8003 : missing source filename

Comment (llunak) [reply src/argprocessing.cpp:250]: But fair enough with clang-cl, I've added setting the slash back when the code gets to treating it as a file.

Comment (rvogg) [reply src/argprocessing.cpp:250]: Ok I didn't try it, I just looked it up in the Microsoft doc. (https://docs.microsoft.com/en-us/cpp/build/reference/cl-filename-syntax?view=msvc-160)
But I think it is very dangerous to just change the character here. Even if it does not work currently. 

What if the path is in quotes? Ccache should then receive the path without quotes.

**Comment by jrosdahl** [comment]
I'd appreciate it if you could submit _"handle newlines on Windows properly"_ and _"handle properly also stdout in addition to stderr"_ as separate PRs since they are independent.

**Comment by llunak** [comment]
> I'd appreciate it if you could submit _"handle newlines on Windows properly"_ and _"handle properly also stdout in addition to stderr"_ as separate PRs since they are independent.

They are not, they both modify the block in ccache.cpp around line 950.

**Comment by jrosdahl** [comment]
> They are not, they both modify the block in ccache.cpp around line 950.

I did not mean that the commits as currently written are independent, I meant that the fix and the feature are logically independent of "MSVC fixes". The first can be introduced separately to fix `send_to_stderr` on Windows. This point is quite minor and I'm OK with not doing that one separately if you think it's hard or tedious to do so. The second one however is a generic "we can now cache stdout for any compiler" feature, not an MSVC fix per se.

**Comment by rvogg** [comment]
The commit 60a6d2d51bcd6fb6e618c1be7aba933ae08d0a06 breaks the correct behavior of ccache. 
With `export CCACHE_NOCPP2=1` the preprocessor output is written to stdout.

**Comment by llunak** [comment]
> I did not mean that the commits as currently written are independent, I meant that the fix and the feature are logically independent of "MSVC fixes". The first can be introduced separately to fix `send_to_stderr` on Windows. This point is quite minor and I'm OK with not doing that one separately if you think it's hard or tedious to do so. The second one however is a generic "we can now cache stdout for any compiler" feature, not an MSVC fix per se.

Done in #962.

> The commit [60a6d2d](https://github.com/ccache/ccache/commit/60a6d2d51bcd6fb6e618c1be7aba933ae08d0a06) breaks the correct behavior of ccache. With `export CCACHE_NOCPP2=1` the preprocessor output is written to stdout.

Fixed there.

**Comment by llunak** [comment]
I've separated these commits into more PRs, so this one is now only one commit.

It makes the code properly handle both -option and /option way of writing MSVC options, which also fixes several places where they weren't both listed properly. It is done by converting all /option arguments to -option (rather than listing both cases everywhere and risking mistakes), and if the handling gets to passing the argument as a file it gets changed back.

**Pull request merged**